### PR TITLE
Elastic Defend arm64 is not supported on Windows yet

### PR DESCRIPTION
**Pull request by intxgo**
## What does this PR do?

This PR enhances elastic-endpoint spec to gracefully refrain from ever installing elastic-endpoint on Windows ARM, because Elastic Defend is not supported on it.

The win32 API used `IsWow64Process2` is available only on Windows >= Win10, however Agent is currently compiled using Go 1.21 which requires at least Windows 10 or Server 2016.

The native architecture constants can be found [here on MSDN web](https://learn.microsoft.com/en-us/windows/win32/sysinfo/image-file-machine-constants), however I believe it's unimportant to port them all since Elastic Defend is compiled only for x86_64 for Windows.

Reading environment variables could be an alternative solution, as Windows sets them by default, examples:
``` 
PROCESSOR_ARCHITECTURE=ARM64
PROCESSOR_IDENTIFIER=ARMv8 (64-bit) Family 8 Model 0 Revision   0, QEMU
```

```
PROCESSOR_ARCHITECTURE=AMD64
PROCESSOR_IDENTIFIER=Intel64 Family 6 Model 151 Stepping 2, GenuineIntel
```

```
PROCESSOR_ARCHITECTURE=AMD64
PROCESSOR_IDENTIFIER=AMD64 Family 25 Model 97 Stepping 2, AuthenticAMD
```
However environment variables could be easily modified, therefore calling native API is better solution.

## Why is it important?

Elastic Defend is not provided yet for Windows ARM (Microsoft Surface X or VM on Apple M1/M2/M3) but since Windows Elastic Agent amd64 works fine on Windows ARM in emulation people install it.

Handling gracefully the incompatibility enables installing Agent across a fleet of Windows AMD64 and ARM64 devices with the same policy.

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] I have added an integration test or an E2E test

## Author's Checklist

- [ ]

## Related issues

- 

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

**Review by nfritts — APPROVED** [review]
This looks good to me from Endpoints perspective

File: internal/pkg/agent/application/info/inject_config_windows.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	var kernel32 = syscall.NewLazyDLL("kernel32.dll")
+	var isWow64Process2 = kernel32.NewProc("IsWow64Process2")
+
+	var currentProcessHandle, _ = syscall.GetCurrentProcess()
+	_, _, _ = isWow64Process2.Call(uintptr(currentProcessHandle), uintptr(unsafe.Pointer(&processMachine)), uintptr(unsafe.Pointer(&nativeMachine)))
+
+	var nativeMachineStr string
+
+	switch nativeMachine {
+	case 0x8664:
```
Comment (pchila) [inline internal/pkg/agent/application/info/inject_config_windows.go:77]: Nitpick: it would be nice to define constants for values corresponding to "amd64" and "arm64".
We could use the same names as https://learn.microsoft.com/en-us/windows/win32/sysinfo/image-file-machine-constants (even better if we put a comment with a link to the docs)

**Review by pchila — APPROVED** [review]
Aside from the small nitpick, the change looks good to me

File: internal/pkg/agent/application/info/inject_config_windows.go (hunk 1 above)
Comment (intxgo) [reply internal/pkg/agent/application/info/inject_config_windows.go]: fixed

**Comment by intxgo** [comment]
I'm going to look at the unit test failure

File: internal/pkg/agent/application/info/inject_config_windows.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+			"config": paths.Config(),
+			"home":   paths.Home(),
+			"logs":   paths.Logs(),
+		},
+		"host": map[string]interface{}{
+			"id": hostInfo.Info().UniqueID,
+		},
+		"runtime.os":             runtime.GOOS,
+		"runtime.arch":           runtime.GOARCH,
+		"runtime.native_arch":    nativeArch(),
```
Comment (cmacknz) [inline internal/pkg/agent/application/info/inject_config_windows.go:55]: Rather than duplicating the entire map here, you should be able to call the non-windows (or lets say the common) version and then just add the single additional `runtime.native_arch` key.

For example:

```go
globalCfg, err := agentGlobalConfig()
if err != nil {
    return fmt.Errorf("getting common global config: %w", err)
}

globalCfg["runtime.native_arch"] = nativeArch()

return globalCfg
```

Essentially just augmenting the common configuration with the windows key, rather than duplicating it for both cases.

File: internal/pkg/agent/application/info/inject_config_windows.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	}, nil
+}
+
+func nativeArch() string {
+	var processMachine, nativeMachine uint16
+
+	var kernel32 = syscall.NewLazyDLL("kernel32.dll")
+	var isWow64Process2 = kernel32.NewProc("IsWow64Process2")
+
+	var currentProcessHandle, _ = syscall.GetCurrentProcess()
```
Comment (cmacknz) [inline internal/pkg/agent/application/info/inject_config_windows.go:71]: Do you need to close this? We recently had a severe memory leak from forgetting to close syscall handles on Windows so this is a sensitive subject :)

File: internal/pkg/agent/application/info/inject_config_windows.go (hunk 2 above)
Comment (intxgo) [reply internal/pkg/agent/application/info/inject_config_windows.go:55]: thanks, indeed. I'm also now wondering if this should rather be added to the go-sysinfo package

Comment (cmacknz) [reply internal/pkg/agent/application/info/inject_config_windows.go:55]: I can't think of a reason why you couldn't put it in go-sysinfo, that will allow us to reuse this outside the agent.

Easiest thing to do is open a PR there.

Comment (intxgo) [reply internal/pkg/agent/application/info/inject_config_windows.go]: I've found out that go-sysinfo is only extending golang.org/x/sys/windows, where the latter already includes `IsWow64Process2` so I use it directly instead of loading the dll.

File: internal/pkg/agent/application/info/svc_windows.go [hunk 4]
```diff
@@ -21,9 +23,10 @@ func RunningUnderSupervisor() bool {
 	if err != nil {
 		return false
 	}
-	defer windows.FreeSid(serviceSid)
+	defer func() { _ = windows.FreeSid(serviceSid) }()
```
Comment (intxgo) [inline internal/pkg/agent/application/info/svc_windows.go:26]: unrelated change, fixing lint error: Error return value of `windows.FreeSid` is not checked (errcheck)

File: internal/pkg/agent/application/info/svc_windows.go [hunk 5]
```diff
@@ -21,9 +23,10 @@ func RunningUnderSupervisor() bool {
...
 
-	t, err := windows.OpenCurrentProcessToken()
+	var t windows.Token
+	err = windows.OpenProcessToken(windows.CurrentProcess(), windows.TOKEN_QUERY|windows.TOKEN_QUERY_SOURCE, &t)
```
Comment (intxgo) [inline internal/pkg/agent/application/info/svc_windows.go:29]: unrelated change, fixing lint error: SA1019: windows.OpenCurrentProcessToken is deprecated: Explicitly call OpenProcessToken(CurrentProcess(), ...) with the desired access instead, or use GetCurrentProcessToken for a TOKEN_QUERY token.  (staticcheck)

File: internal/pkg/agent/application/info/svc_windows.go [hunk 6]
```diff
@@ -21,9 +23,10 @@ func RunningUnderSupervisor() bool {
 	if err != nil {
 		return false
 	}
-	defer windows.FreeSid(serviceSid)
+	defer func() { _ = windows.FreeSid(serviceSid) }()
 
-	t, err := windows.OpenCurrentProcessToken()
+	var t windows.Token
+	err = windows.OpenProcessToken(windows.CurrentProcess(), windows.TOKEN_QUERY, &t)
```
Comment (intxgo) [inline internal/pkg/agent/application/info/svc_windows.go:29]: unrelated change, fixing lint error: SA1019: windows.OpenCurrentProcessToken is deprecated: Explicitly call OpenProcessToken(CurrentProcess(), ...) with the desired access instead, or use GetCurrentProcessToken for a TOKEN_QUERY token. (staticcheck)

**Comment by intxgo** [comment]
could you review this again since the implementation has changed significantly

@michalpristas @pchila 

The unit test error on Windows Server 2016 (`Failed to find IsWow64Process2 procedure in kernel32.dll: The specified procedure could not be found.`) will be resolved by `go-sysinfo` update, as soon as [the improvement](https://github.com/elastic/go-sysinfo/pull/201) is merged

**Review by michalpristas — APPROVED** [review]
i think it looks good, fix from sysinfo is included for older windows versions as well.
haven't tested this on arm machine

**Comment by cmacknz** [comment]
The one test failure here is unrelated:

```
	Received unexpected error:
        	            	failed to download https://snapshots.elastic.co/8.13.0-e2cda7bd/downloads/beats/elastic-agent/elastic-agent-8.13.0-SNAPSHOT-linux-x86_64.tar.gz: failed to write file /home/ubuntu/agent/.agent-testing/artifact/elastic-agent-8.13.0-SNAPSHOT-linux-x86_64.tar.gz: stream error: stream ID 1; INTERNAL_ERROR; received from peer
```

**Comment by cmacknz** [comment]
Going to force merge this one.
